File: src/main/java/no/nav/fo/veilarbregistrering/enhet/adapter/EnhetRestClient.java
package no.nav.fo.veilarbregistrering.enhet.adapter;

import com.google.gson.*;
import no.nav.common.rest.client.RestClient;
import no.nav.common.rest.client.RestUtils;
import no.nav.fo.veilarbregistrering.arbeidsforhold.Organisasjonsnummer;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.ws.rs.NotFoundException;
import java.io.IOException;
import java.lang.reflect.Type;
import java.time.LocalDate;
import java.util.Optional;

import static java.util.concurrent.TimeUnit.MILLISECONDS;


class EnhetRestClient {

    private static final long HTTP_READ_TIMEOUT = 120000;

    private static final Logger LOG = LoggerFactory.getLogger(EnhetRestClient.class);

    private static final Gson gson = new GsonBuilder().registerTypeAdapter(LocalDate.class, new LocalDateDeserializer()).create();

    private final String url;
    private final OkHttpClient client;

    EnhetRestClient(String baseUrl) {
        this.url = baseUrl + "/v1/organisasjon/";
        client = RestClient.baseClientBuilder().readTimeout(HTTP_READ_TIMEOUT, MILLISECONDS).build();
    }

    Optional<OrganisasjonDetaljerDto> hentOrganisasjon(Organisasjonsnummer organisasjonsnummer) {
        try {
            String response = utfoerRequest(organisasjonsnummer);
            OrganisasjonDto organisasjonDto = parse(response);
            return Optional.of(organisasjonDto.getOrganisasjonDetaljer());

        } catch (NotFoundException e) {
            LOG.warn("Fant ikke organisasjon for organisasjonsnummer", e);
            return Optional.empty();

        } catch (RuntimeException e) {
            throw new RuntimeException("Hent organisasjon feilet", e);
        }
    }

    String utfoerRequest(Organisasjonsnummer organisasjonsnummer) {
        Request request = new Request.Builder()
                .url(this.url + organisasjonsnummer.asString())
                .build();
        try {
            Response response = client.newCall(request).execute();
            RestUtils.throwIfNotSuccessful(response);
            return RestUtils.getBodyStr(response).orElseThrow(RuntimeException::new);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    static OrganisasjonDto parse(String jsonResponse) {
        return gson.fromJson(jsonResponse, OrganisasjonDto.class);
    }

    private static class LocalDateDeserializer implements JsonDeserializer<LocalDate> {
        public LocalDate deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
                throws JsonParseException {

            return Optional.ofNullable(json.getAsJsonPrimitive().getAsString())
                    .map(LocalDate::parse)
                    .orElse(null);
        }
    }
}

File: src/main/kotlin/no/nav/fo/veilarbregistrering/enhet/adapter/EnhetRestClient.kt
package no.nav.fo.veilarbregistrering.enhet.adapter

import no.nav.common.health.HealthCheck
import no.nav.common.health.HealthCheckResult
import no.nav.common.health.HealthCheckUtils
import no.nav.common.rest.client.RestClient
import no.nav.common.rest.client.RestUtils
import no.nav.common.utils.UrlUtils
import no.nav.fo.veilarbregistrering.arbeidsforhold.Organisasjonsnummer
import no.nav.fo.veilarbregistrering.config.objectMapper
import no.nav.fo.veilarbregistrering.log.loggerFor
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.IOException
import java.util.concurrent.TimeUnit

open class EnhetRestClient(val baseUrl: String): HealthCheck {
    private val url: String = "$baseUrl/api/v1/organisasjon/"

    open fun hentOrganisasjon(organisasjonsnummer: Organisasjonsnummer): OrganisasjonDetaljerDto? {
        val request = Request.Builder()
                .url(url + organisasjonsnummer.asString())
                .build()

        return try {
            client.newCall(request).execute().use { response ->
                when (response.code()) {
                    in 200..399 -> {
                        val body = RestUtils.getBodyStr(response).orElseThrow { RuntimeException() }
                        parse(body).organisasjonDetaljer
                    }
                    404 -> {
                        LOG.warn("Fant ikke organisasjon for organisasjonsnummer: $organisasjonsnummer")
                        null
                    }
                    else -> throw RuntimeException("Hent organisasjon feilet med status: ${response.code()}")
                }
            }
        } catch (e: IOException) {
            throw RuntimeException("Hent organsisasjon feilet", e)
        }
    }

    override fun checkHealth(): HealthCheckResult {
        return HealthCheckUtils.pingUrl(UrlUtils.joinPaths(baseUrl, "/api/ping"), RestClient.baseClient())
    }

    companion object {
        private const val HTTP_READ_TIMEOUT: Long = 120000
        private val LOG = loggerFor<EnhetRestClient>()
        private val client: OkHttpClient =
                RestClient.baseClientBuilder().readTimeout(HTTP_READ_TIMEOUT, TimeUnit.MILLISECONDS).build()

        fun parse(jsonResponse: String): OrganisasjonDto {
            return objectMapper.readValue(jsonResponse, OrganisasjonDto::class.java)
        }
    }
}